File: Contests/Codechef/Contest/Long Challenges/06_15/MOREFB.py
from sys import exit,stdin
from collections import Counter
from time import time

class MatrixFibonacci:
	Q = [[1, 1],
		 [1, 0]]

	def __init__(self):
		self.__memo = {}

	def __multiply_matrices(self, M1, M2):
		a11 = M1[0][0]*M2[0][0] + M1[0][1]*M2[1][0]
		a12 = M1[0][0]*M2[0][1] + M1[0][1]*M2[1][1]
		a21 = M1[1][0]*M2[0][0] + M1[1][1]*M2[1][0]
		a22 = M1[1][0]*M2[0][1] + M1[1][1]*M2[1][1]
		r = [[a11, a12], [a21, a22]]
		return r

	def __get_matrix_power(self, M, p):
		
		if p == 1:
			return M
		if p in self.__memo:
			return self.__memo[p]
		K = self.__get_matrix_power(M, int(p/2))
		R = self.__multiply_matrices(K, K)
		self.__memo[p] = R
		return R

	def get_number(self, n):
		if n == 0:
			return 0
		if n == 1:
			return 1
		powers = [int(pow(2, b))
				  for (b, d) in enumerate(reversed(bin(n-1)[2:])) if d == '1']
		matrices = [self.__get_matrix_power(MatrixFibonacci.Q, p)
					for p in powers]
		while len(matrices) > 1:
			M1 = matrices.pop()
			M2 = matrices.pop()
			R = self.__multiply_matrices(M1, M2)
			matrices.append(R)
		return matrices[0][0][0]
def subset(arr,size,left, index,l):
	if(left==0):
		d[sum(l)]+=1
		return
	
	for i in xrange(index,size):
		l.append(arr[i])
		subset(arr,size,left-1,i+1,l)
		l.pop() 
d=Counter()
def main():
	MOD=99991
	f_sum=0
	mfib =MatrixFibonacci()
	n,k=map(int,stdin.readline().split())
	st=map(int,stdin.readline().split())
	#start=time()
	lt=list()   
	subset(st,n,k,0,lt)

	for i in d:
		f_sum=(f_sum+d[i]*mfib.get_number(i))

	
	print f_sum%MOD
	#print time()-start
	exit(0)


if __name__=="__main__":
	main()

File: Contests/Codechef/Contest/Long Challenges/06_15/MOREFB.cpp
#include <bits/stdc++.h>
#define FOR(i,j,k) for(int i=j;i<k;i++)
#define ll long long
#define ui unsigned int
#define DEBUG(x) cout << '>' << #x << ':' << x << endl;
#define MOD 99991
using namespace std;
map<ll, ll > d;
map<ll, ll>::iterator iter;

ll fib(ll n)
{
  ll a = 0, b = 1, c, i;
  if( n == 0)
    return a;
  for (i = 2; i <= n; i++)
  {

     c = a%MOD + b%MOD;
     a = b%MOD;
     b = c;

  }
  return b%MOD;
}


void subset(vector<ll> arr, ll size, ll left, ll index, list<ll> &l){
    if(left==0){
        ll sum=0;
        for(list<ll>::iterator it=l.begin(); it!=l.end() ; ++it){
            sum+=*it;
        }
        if(d.count(sum))
            d[sum]+=1;
        else
            d[sum]=1;
        return;
    }
    for(ll i=index; i<size;i++){
        l.push_back(arr[i]);
        subset(arr,size,left-1,i+1,l);
        l.pop_back();   
    }

}     

int main(){

    list<ll> lt;
    vector<ll> array;  
    ll f_sum=0,n,k,tmp;
    cin>>n>>k;
    FOR(i,0,n){
        cin>>tmp;
        array.push_back(tmp);
    }
    
    subset(array,n,k,0,lt);
    for(iter = d.begin(); iter != d.end(); iter++){
     f_sum+=( (((*iter).second)%MOD)*(fib((*iter).first)%MOD) )%MOD;
     if(f_sum>MOD)
        f_sum %=MOD;
    }
    cout<<f_sum;

    return 0;
}